Modify nodes to register directly with the master.

roberthbailey opened the pull request:
Ref #6087. 
Fixes #7495.

Commit:
Modify nodes to register directly with the master.

 - Delete nodes when they are no longer ready and don't exist in the
cloud provider.
 - Label each node with it's hostname.
 - Add flag to skip node registration.
 - Add a test for registering an existing node.

roberthbailey commented:
/cc @bgrant0607

roberthbailey commented on pkg/tools/etcd_helper.go:
```diff
@@ -165,12 +165,14 @@ func (h *EtcdHelper) ExtractObjToList(key string, listObj runtime.Object) error
 
 	nodes := make([]*etcd.Node, 0)
-	nodes = append(nodes, response.Node)
+	if !IsEtcdNotFound(err) {
```
@smarterclayton I found that if there is an etcd not found error here (type 100) then response will be null so response.Node or response.EtcdIndex (below) will cause this goroutine to panic.

smarterclayton replied:
Derek just fixed this in another pull.

smarterclayton replied:
https://github.com/GoogleCloudPlatform/kubernetes/pull/6938 covers it

> On Apr 16, 2015, at 9:30 PM, Robert Bailey <notifications@github.com> wrote:
> 
> In pkg/tools/etcd_helper.go:
> 
> > @@ -165,12 +165,14 @@ func (h *EtcdHelper) ExtractObjToList(key string, listObj runtime.Object) error
> >  	}
> >  
> >  	nodes := make([]*etcd.Node, 0)
> > -	nodes = append(nodes, response.Node)
> > +	if !IsEtcdNotFound(err) {
> @smarterclayton I found that if there is an etcd not found error here (type 100) then response will be null so response.Node or response.EtcdIndex (below) will cause this goroutine to panic.
> 
> —
> Reply to this email directly or view it on GitHub.
>

roberthbailey replied:
I've rebased on top of that PR.

roberthbailey commented:
@derekwaynecarr @pires @justinsb I haven't been able to test this on other cloud providers -- can you take a look and see if you think it will break anything?

smarterclayton commented:
What are the security implications of this?  How do I control as an admin clients interfering with other nodes?  Does this require the cloud provider on the kubelet to have security credentials to the cluster (very scary)?  Can I still disable the cloud provider on the kubelet?

roberthbailey commented:
Security implications: A node with credentials can add itself to the cluster. Right now, nodes can do this because they have bearer tokens. 

Right now we don't have any authz in the apiserver. Once we do, we should add policies that only allow nodes to update their own data (and not that of other nodes). Any node today has free reign to update _anything_ in the APIserver because it has full read/write access through the secured port with the node bearer token. 

As a first cut I just moved the cloud provider bits from the nodecontroller to the kubelet. But I think we should be able to remove most/all of them. For GCE (and likely AWS) we may still want make a call into the cloud provider to hit the local metadata server and get the external IP address of a node. But calling into the cloud provider to determine system resources seems silly. 

Yes, you can still disable the cloud provider on the kubelet. When you run `./hack/test-cmd.sh` it runs the kubelet without a cloud provider (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/hack/test-cmd.sh#L72-L83) and all tests still pass.

smarterclayton commented:
> On Apr 16, 2015, at 9:50 PM, Robert Bailey <notifications@github.com> wrote:
> 
> Security implications: A node with credentials can add itself to the cluster. Right now, nodes can do this because they have bearer tokens.
> 
> Right now we don't have any authz in the apiserver.
> 
The policy model @deads2k has created for OpenShift (which post 1.0 we'll evaluate bringing into kube) should be able to handle this.  We also have the pattern to properly give kubelets unique identity, although we do not yet have the ideal policies.
> Once we do, we should add policies that only allow nodes to update their own data (and not that of other nodes). Any node today has free reign to update anything in the APIserver because it has full read/write access through the secured port with the node bearer token.
> 
I just wanted to make sure this didn't change the detente.  Thanks.
> As a first cut I just moved the cloud provider bits from the nodecontroller to the kubelet. But I think we should be able to remove most/all of them. For GCE (and likely AWS) we may still want make a call into the cloud provider to hit the local metadata server and get the external IP address of a node. But calling into the cloud provider to determine system resources seems silly.
> 
> Yes, you can still disable the cloud provider on the kubelet. When you run ./hack/test-cmd.sh it runs the kubelet without a cloud provider (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/hack/test-cmd.sh#L72-L83) and all tests still pass.
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
 		return
 	}
+	if err := kl.registerNode(); err != nil {
+		glog.Errorf("Failed to register node: %s. Giving up.", err)
```
We should use util.HandleError here

roberthbailey replied:
I didn't realize that existed -- cool. I've passed it a new error so that we keep the custom text on top of the error text returned from registerNode.

pires commented on pkg/kubelet/kubelet.go:
```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
+}
+
+// registerNode registers the node with the cluster master.
+func (kl *Kubelet) registerNode() error {
```
Will this replace what `kube-register` does?

roberthbailey replied:
This change will make clustering much, much simpler. I haven't looked at kube-register in detail, but this may make it unnecessary because as long as you point kubelets at the master they will join the cluster and become schedulable.

kelseyhightower replied:
This would be so nice. I only wrote kube-register to solve this very problem.

pires commented:
@roberthbailey right now I can only test on GCE since I no longer use AWS.

Now, from what I understand from your changes, you're moving cloud-provider stuff from `controller-manager` (poll) to `kubelet` (push). If my assumption is correct, I think this is great because it can replace what I and others (cc @AntonioMeireles @wkharold) already use with [`kube-register`](https://github.com/kelseyhightower/kube-register).

After reading @smarterclayton I too believe there are risks, but right now, as a sysadmin, I'm OK with it.

roberthbailey commented:
@pires I'm trying to move the node creation from being done in a non-obvious way (a control loop that talks to the cloud provider and uses a regular expression to decide which nodes should be part of the cluster) to an explicit call from a node to join a specific master. 

Right now this call "just works" because the kubelet already has credentials to talk to the master, but in the future (with [dynamic clustering](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/clustering.md#dynamic-clustering)) the node will first ask to join by sending a CSR to the master to get itself credentials to POST and create a node entry. This will give sysadmins the ability to set policies on whether nodes can join automatically or whether they need to be manually approved to join a cluster. But that's post-1.0. For now, if a node has credentials, it will automatically join the cluster. 

Some of the fields for the node may still make sense to fill in from the node controller (which is why I cc'd @bgrant0607 as I thought he'd have an opinion here) but I think that most of the ones we are setting are better known by the kubelet itself (e.g. node resources, hostname, ip addresses).

pires commented:
@roberthbailey just to be sure, this would remove the need for tools such as `kube-register`, right?

roberthbailey commented:
@pires I think so. Would it be possible for you to test this CL on CoreOS and find out?

AntonioMeireles commented:
@roberthbailey given the nature of this @kelseyhightower and his pals at CoreOS are the most abilitated to postulate about this. anyway i *may* be able to test this CL early next week...

pires commented:
@roberthbailey sure will you be available on IRC for debugging?

roberthbailey commented:
I'll hop on right now.

pires commented:
Removed `kube-register` from master node, replaced `kubelet` with @roberthbailey version and... it works!

```
$ kubectl get minions
NAME           LABELS        STATUS
172.17.8.102   Schedulable   <none>    Ready
172.17.8.103   Schedulable   <none>    Ready
```

bgrant0607 commented:
cc @gmarek @davidopp

bgrant0607 commented:
Nodes shouldn't need to call the cloud provider.

We have discussed making it possible to contact the local metadata server for clouds that provide that, but I'd like to avoid that, since it's not universally available.

We need to make some changes to the Node object. The node name is currently overloaded for at least 2 purposes: object name and cloud provider ("external") ID. It's often set to the hostname or address, which adds to the confusion about the name requirements and usage.

The only information currently required to create a node is the externalID, which is automatically defaulted to the node name if not provided -- that part I don't mind. However, I'd like the externalID to be used to lookup the node with the cloud provider, rather than looking up the external ID using the node name:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/cloudprovider/nodecontroller/nodecontroller.go#L640

Addresses are also looked up via the cloud provider. It's true that the node knows at least a subset of its addresses, though not necessarily external ones. The node controller could fill in additional addresses after creation (late initialization).

We don't really have a good way of knowing which address is preferred for master-to-node communication (/validate, log access, exec, port forwarding, proxy, redirect), and we don't know which name/address is specified in the node cert, to the extent that's relevant -- #6985. 

Some master component will allocate the PodCIDR via late initialization.

Capacity can be populated after creation, also.

bgrant0607 commented:
This should also allow us to get rid of the --machines flag, which @eparis should also like.

kube-register allows users to specify fleet node labels to select which nodes to add to Kubernetes. That's still useful, so I don't think this obsoletes kube-register. I wouldn't mind adding fleet/kube-register as a cloudprovider -- #2890 @kelseyhightower .

bgrant0607 commented:
So maybe the one cloudprovider bit the node needs is the ability to get its externalID, which it should get from a local metadata server, or maybe just use the hostname if no cloudprovider.

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
+		ObjectMeta: api.ObjectMeta{Name: kl.hostname},
+		Status: api.NodeStatus{
+			Capacity: api.ResourceList{
+				api.ResourceCPU:    *resource.NewMilliQuantity(1000, resource.DecimalSI),
```
Kubelet already has code for populating this with real values.
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1768

roberthbailey replied:
Why ask the cloud provider to determine the node resources when the node can just self report them? How will that work for deployments without a cloud provider?

Also, see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/cloudprovider/gce/gce.go#L512

bgrant0607 replied:
No good reason -- legacy cruft. We must get the values from the node.

davidopp replied:
I think this code was in NodeController so there would be some capacity in NodeStatus between the time the node is created in the master, and the first time the node reports its status.

erictune replied:
Fine with leaving it this way in this PR, but we should clean up at some point.  Please leave a TODO if you don't think you can get to it soon.

roberthbailey commented:
> kube-register allows users to specify fleet node labels to select which nodes to add to Kubernetes.

Could this be done instead by using fleet node labels to choose which --api_server argument we pass to the kubelet on each node? It doesn't make sense for a kubelet to connect to an apiserver and watch for pods to run (and send events) if it isn't actually part of the cluster.

bgrant0607 commented on pkg/cloudprovider/nodecontroller/nodecontroller.go:
```diff
@@ -20,11 +20,9 @@ import (
-	apierrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/record"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider"
```
You should be able to delete the cloudprovider here.

roberthbailey replied:
Now that I re-added the code to handle deleting nodes that no longer exist, we are still using the cloud provider.

bgrant0607 commented:
Actually, I'm liking this more. I think we should eliminate cloudprovider stuff from the nodecontroller and move it out of the cloudprovider subtree, as discussed in #4851.

The node should get any info it needs from the metadata server, if present, or from a local config file, fleet, whatever -- the mechanism should be pluggable. Calling the existing cloudprovider API could be a stopgap.

We'll likely need to resurrect a cloudprovider-oriented nodecontroller in the future in order to automatically repair and provision nodes (cluster auto-scaling), but that's ok.

bgrant0607 commented:
@roberthbailey Fair point about using fleet to set the api_server flag (though I wish it were from a config file instead of a flag).

bgrant0607 commented on cluster/saltbase/salt/kubelet/default:
```diff
@@ -15,6 +15,15 @@
   {% set api_servers = "--api_servers=https://" + ips[0][0] + ":6443" -%}
 {% endif -%}
 
+# Disable registration for the kubelet running on the master on GCE.
```
Eventually we do want to register the master kubelet. Registration just needs to be retried if the apiserver doesn't respond. And, for now, we want scheduling to be disabled for the master node. #6428
cc @dchen1107 @vishh

vishh replied:
Why disable registration of the master kubelet? The API server reports the master pods but the usual APIs used to access the pods, like '/containerLogs', '/exec', etc, are not working as of now.

roberthbailey replied:
It keeps things consistent with the current deployment. I'd be happy to change this in a future PR to be configurable, but didn't want too many moving pieces here. For one, we'd need to change cluster validation to expect N+1 nodes (rather than N) since the master is now counted. And I didn't want to worry about other ramifications at the same time as changing the node creation pattern.

wkharold replied:
Robert, fyi I tested your kubelet mod in my GCE environment this weekend.
Worked perfectly! I deploy nodes with cloud configs derived from Paulo's
and the nodes show up with proper names that I can then use to create
services with an external load balancer.

Very nice, Thanks ... WkH

On Fri, Apr 17, 2015 at 8:19 PM, Robert Bailey <notifications@github.com>
wrote:

> In cluster/saltbase/salt/kubelet/default
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#discussion_r28639946>
> :
>
> > @@ -15,6 +15,15 @@
> >    {% set api_servers = "--api_servers=https://" + ips[0][0] + ":6443" -%}
> >  {% endif -%}
> >
> > +# Disable registration for the kubelet running on the master on GCE.
>
> It keeps things consistent with the current deployment. I'd be happy to
> change this in a future PR to be configurable, but didn't want too many
> moving pieces here. For one, we'd need to change cluster validation to
> expect N+1 nodes (rather than N) since the master is now counted. And I
> didn't want to worry about other ramifications at the same time as changing
> the node creation pattern.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r28639946>
> .
>



-- 
... WkH

pires replied:
This is so awesome Ward! Can't wait to see this merged.
On Apr 19, 2015 6:57 PM, "wkharold" <notifications@github.com> wrote:

> In cluster/saltbase/salt/kubelet/default
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#discussion_r28657993>
> :
>
> > @@ -15,6 +15,15 @@
> >    {% set api_servers = "--api_servers=https://" + ips[0][0] + ":6443" -%}
> >  {% endif -%}
> >
> > +# Disable registration for the kubelet running on the master on GCE.
>
> Robert, fyi I tested your kubelet mod in my GCE environment this weekend.
> Worked perfectly! I deploy nodes with cloud configs derived from Paulo's
> and the nodes show up with proper names that I can then use to create
> services with an external load balancer. Very nice, Thanks ... WkH
> … <#14cd3e667ba8e637_>
> On Fri, Apr 17, 2015 at 8:19 PM, Robert Bailey <notifications@github.com>
> wrote: In cluster/saltbase/salt/kubelet/default <
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#discussion_r28639946>
> : > @@ -15,6 +15,15 @@ > {% set api_servers = "--api_servers=https://" +
> ips[0][0] + ":6443" -%} > {% endif -%} > > +# Disable registration for the
> kubelet running on the master on GCE. It keeps things consistent with the
> current deployment. I'd be happy to change this in a future PR to be
> configurable, but didn't want too many moving pieces here. For one, we'd
> need to change cluster validation to expect N+1 nodes (rather than N) since
> the master is now counted. And I didn't want to worry about other
> ramifications at the same time as changing the node creation pattern. —
> Reply to this email directly or view it on GitHub <
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r28639946>
> .
>  -- ... WkH
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r28657993>
> .
>

erictune replied:
Fine with you doing the master registration in a future PR, as long as it gets done at some point (with appropriate adjustments for GKE)

zmerlynn commented:
> The node should get any info it needs from the metadata server, if present, or from a local config file, fleet, whatever -- the mechanism should be pluggable. Calling the existing cloudprovider API could be a stopgap.

I'm missing something. This sounds like it's asking for an interface of some sort - it may not be the existing contents of `cloudprovider/`, but it sure sounds like a library of some sort?

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
+		if !ok {
+			return nil, fmt.Errorf("failed to get instances from cloud provider")
+		}
+		// TODO(roberthbailey): Inspect the local machine rather than calling out
```
Yeah, just delete this.

bgrant0607 commented on pkg/cloudprovider/nodecontroller/nodecontroller.go:
```diff
@@ -187,198 +144,6 @@ func (nc *NodeController) Run(period time.Duration, syncNodeList bool) {
-
-// reconcileExternalServices updates balancers for external services, so that they will match the nodes given.
-// Returns true if something went wrong and we should call reconcile again.
-func (nc *NodeController) reconcileExternalServices(nodes *api.NodeList) (shouldRetry bool) {
```
Is this done by the new servicecontroller now?

roberthbailey replied:
I filed #6926 to move it there, but haven't done so yet. I figured @a-robinson would get it moved over before we had agreement on this PR and it was ready to be merged. Otherwise I'll move it before this is merged.

bgrant0607 replied:
It doesn't seem to call UpdateTCPLoadBalancer.

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
+			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = instanceID
+		nodeAddresses, err := instances.NodeAddresses(kl.hostname)
```
This isn't doing everything that populateAddresses was doing. 

Is the legacy hostIP populated? We need that until we rip out v1beta1/2.

We also should decide whether setting the addresses at registration time is sufficient, or whether we need to sync them periodically. Is it possible to add external addresses after creation on any cloudprovider?

roberthbailey replied:
I thought I copied this directly from populateAddresses but it sounds like I missed a bit. I'll go back and check. 

Addresses are part of the node status. So it should be possible for the node to notice that it's address changed and send that as part of the regular status update to the master (It doesn't do that yet).

davidopp replied:
For some background, see #6072. Currently we handle updating IP address when node is added (or node deleted and then re-added) but not when node changes IP address on the fly.

smarterclayton commented:
Does that fundamentally make it different than cloud provider then?  Just "node machine id provider"?  

> On Apr 17, 2015, at 7:41 PM, Brian Grant <notifications@github.com> wrote:
> 
> So maybe the one cloudprovider bit the node needs is the ability to get its externalID, which it should get from a local metadata server, or maybe just use the hostname if no cloudprovider.
> 
> —
> Reply to this email directly or view it on GitHub.
>

zmerlynn commented:
@smarterclayton: It's still a node in a cloud.

zmerlynn commented:
... or, rather, that may have been provisioned by a cloud provider. Or not.

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
+			continue
+		}
+		glog.V(2).Infof("Attempting to register node %s", node.Name)
+		if _, err := kl.kubeClient.Nodes().Create(node); err != nil {
```
Does this correctly handle the case where the node is already registered?

roberthbailey replied:
No it doesn't. I thought of that in the shower this morning and was discussing it with @zmerlynn in regards to reboot/update but got sidetracked with meetings and didn't get around to fixing it. The fix shouldn't be difficult. Thanks for the reminder.

roberthbailey replied:
This now handles previously registered nodes. I did a reboot test and verified that the node switches from Ready -> NotReady -> Ready.

bgrant0607 commented:
It would be nice if there were a way for the node to provide labels at registration time, also.

bgrant0607 commented:
What's going to delete nodes in the case that the corresponding VMs have been nuked?

smarterclayton commented:
Just thinking along Brian's lines that the interface for "get machine id" has little to do with cloud, and is likely to mix and match depending on what your clusters look like.

> On Apr 17, 2015, at 8:24 PM, Zach Loafman <notifications@github.com> wrote:
> 
> ... or, rather, that may have been provisioned by a cloud provider. Or not.
> 
> —
> Reply to this email directly or view it on GitHub.
>

roberthbailey commented:
> What's going to delete nodes in the case that the corresponding VMs have been nuked?

They don't get deleted (but I'm not sure they would today either). [monitorNodeStatus](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/cloudprovider/nodecontroller/nodecontroller.go#L527-L529) will still move them into the `ConditionUnknown` state, and we could update that function to eventually garbage collect old nodes or provide a command in kubectl to remove nodes.

davidopp commented:
> They don't get deleted (but I'm not sure they would today either)

Unless I'm mistaken, the following NodeController code today does delete nodes that have been removed from the cloud provider:
https://github.com/GoogleCloudPlatform/kubernetes/blob/65d132a3eecee38fc22e82ff5fd9e100aae342be/pkg/cloudprovider/nodecontroller/nodecontroller.go#L324

erictune commented on pkg/kubelet/kubelet.go:
```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
+// registerNode registers the node with the cluster master.
+func (kl *Kubelet) registerNode() error {
+	step := 100 * time.Millisecond
+	for start := time.Now(); time.Since(start) < 30*time.Minute; time.Sleep(step) {
```
I don't see the purpose of timing out after 30 minutes.

roberthbailey replied:
Do you think it should just keep trying indefinitely?

smarterclayton replied:
Maybe with a back off?  100ms early, every few seconds later?

----- Original Message -----
> > +			return nil, fmt.Errorf("failed to get instance ID from cloud provider:
> > %v", err)
> > +		}
> > +		node.Spec.ExternalID = instanceID
> > +		nodeAddresses, err := instances.NodeAddresses(kl.hostname)
> > +		if err != nil {
> > +			return nil, fmt.Errorf("failed to get node address from cloud provider:
> > %v", err)
> > +		}
> > +		node.Status.Addresses = nodeAddresses
> > +	}
> > +	return node, nil
> > +}
> > +
> > +// registerNode registers the node with the cluster master.
> > +func (kl *Kubelet) registerNode() error {
> > +	step := 100 * time.Millisecond
> > +	for start := time.Now(); time.Since(start) < 30*time.Minute;
> > time.Sleep(step) {
> 
> Do you think it should just keep trying indefinitely?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r28936676

erictune replied:
If you think that exiting and restarting could possibly help, than it should crash after a minute or so, and then be restarted by monit.  

Otherwise, it should retry forever.

When something goes wrong in the cluster, the person debugging it will make changes, and wait a little while to see if they help.  Adding a magic 30minute time constant to that feedback loop doesn't help.

roberthbailey replied:
@smarterclayton The current code implements backoff (but not jitter). Start at 100ms, doubles on each try until it hits 7 seconds, and then stays at 7 seconds until it reaches the timeout (30 minutes). 

@erictune I see your point. Which option do you think is better: crashing or retrying forever?

erictune commented:
LGTM except for outstanding comment about populateAddresses, and needs rebase.

roberthbailey commented:
@erictune I have quite a few local changes (including a rebase) that need to be pushed. I'll ping this PR when it's ready for a re-review.

roberthbailey commented:
I pushed a new commit that addresses most of the comments received thus far. I'm still waiting on #7251 before doing a final test and verifying that I didn't delete any necessary code from nodecontroller.go.

kelseyhightower commented:
This all looks pretty good to me. I only wrote kube-register to automate this process. I'll continue to manage kube-register until this becomes the standard.

kelseyhightower commented:
As @erictune mentioned, the only other value provided by kube-register is the ability to filter nodes based on info stored in fleet. This maybe really useful for people running fleet and want a simple way to selectively add nodes to an k8s cluster.

FYI - kube-register was recently updated to support the v1beta3 API.

bgrant0607 commented:
@kelseyhightower The nodes need to be configured with which apiserver to connect to, so that seems like enough of a selection mechanism.

I would like a way for nodes to discover what labels they should carry (such as from a configurable file or URL), so that they can specify those labels when they register.

roberthbailey commented:
@erictune - This is almost ready, but now I'm seeing cluster validation fail during kube-up. It works if I run it manually after the cluster has been created though. Still investigating...

roberthbailey commented:
Ok, figured it out. If no nodes are registered yet, then `kubectl get nodes` will return an empty list (correctly) but the `grep -c` line was returning a 1 exit code because the file had to matches. Since we have `set -o errexit` at the top of the script this caused kube-up to exit prematurely. 

@erictune Please don't merge this before 0.16.0 is cut tomorrow. I'd rather have it bake in test for a week or so before it's part of an official release.

roberthbailey commented:
For the record, e2e tests all pass:

Ran 35 of 40 Specs in 1055.414 seconds
SUCCESS! -- 35 Passed | 0 Failed | 1 Pending | 4 Skipped I0426 21:44:35.961197   48921 driver.go:96] All tests pass

gmarek commented:
@gmarek - I want to take a look at this, but didn't manage today. Please don't block on me if this needs to be merged quickly.

gmarek commented:
I'm slightly worried that this PR removes some tests and adds none... This is a relatively major change, and I'd rather have tests for it.

Other than that we need to discuss handling PodCIDRs in a short term.

smarterclayton commented:
I talked with some of our ops folks and they were concerned that this is automatic and can't be bypassed.  Can we make this behavior optional?  The primary concerns were that they wanted the administrator to still have the ability to manage this on their own (by adding node resources) and so they'd like to disable this capability in some environments.  We have policy that can be applied to block create node on the apiserver, so having the kubelet try and fail doesn't help.  

Can you add a kubelet flag for the new behavior (I'm ok with it being default true)?

gmarek commented on pkg/kubelet/kubelet.go:
```diff
@@ -1863,25 +1915,54 @@ func (kl *Kubelet) recordNodeUnschedulableEvent() {
+	} else {
+		addr := net.ParseIP(kl.hostname)
+		if addr != nil {
+			node.Status.Addresses = []api.NodeAddress{{Type: api.NodeLegacyHostIP, Address: addr.String()}}
```
I don't think that using LegacyHostIP is a good way to go.

roberthbailey replied:
@gmarek - @bgrant0607 said that we had to keep this behavior in until we remove the v1beta1/v1beta2 apis for backwards compatibility. It should be removed once the older API versions are phased out.

gmarek replied:
IIUC v1beta1 and v1beta2 are in the process of being removed. I'm not quite clear how advanced it is, but #8087 is there.

roberthbailey replied:
I'm going to leave it in in the hope that my PR gets merged before v1beta1 is completely ripped out.

roberthbailey commented:
> Can you add a kubelet flag for the new behavior (I'm ok with it being default true)?

If you don't specify the --api_servers flag then the kubelet won't try to register itself. I think for now this should meet your criteria of turning it off with a flag. If you want to have a pool of stand-by kubelets, you launch them in "free-range" mode (e.g. not attached to a cluster) and then when you want to add one to a cluster you restart it and add the --api_servers flag and it will self-register (assuming it has the correct auth bits in place for the cluster).

roberthbailey commented:
> I'm slightly worried that this PR removes some tests and adds none

That is fair, but the functionality here (clustering) is pretty well tested by the e2e tests. If kubelets don't properly register with the master then the e2e cluster creation will fail and we will notice pretty quickly. We are planning on adding a reboot e2e test soon which will help with a manual test that I did, and once we can dynamically add/remove nodes from a cluster we will want e2e tests for that as well. 

Is there anything in particular that I added that you think warrants unit tests?

smarterclayton commented:
> On Apr 28, 2015, at 10:46 AM, Robert Bailey <notifications@github.com> wrote:
> 
> Can you add a kubelet flag for the new behavior (I'm ok with it being default true)?
> 
> If you don't specify the --api_servers flag then the kubelet won't try to register itself. I think for now this should meet your criteria of turning it off with a flag. If you want to have a pool of stand-by kubelets, you launch them in "free-range" mode (e.g. not attached to a cluster) and then when you want to add one to a cluster you restart it and add the --api_servers flag and it will self-register (assuming it has the correct auth bits in place for the cluster).
> 
I don't think that satisfies our concerns.  Our ops guys feel very strongly that self registration is not the only way that nodes should be registered and need the ability to guarantee that kubelets have to be accepted before they can join the cluster.  This change, as is, appears to require the kubelet to have create node permission (correct me if wrong).  We would prefer the kubelet have a way of not explicitly self registering (or being able to handle being told self registration is forbidden by the apiserver).
> —
> Reply to this email directly or view it on GitHub.
>

roberthbailey commented:
So you want the kubelet to be able to connect to an apiserver (to send events and node status updates) but not to register itself (e.g. the old behavior)? And you'll use something like kube-register to manually create each of the nodes in the cluster? (Just trying to better understand your use case)

Accepting each kubelet before it joins the cluster is part of the design for dynamic clustering (post v1.0)  but is out of scope for static clustering (what this PR implements). Hopefully when we converge on dynamic clustering this becomes much simpler. 

I can certainly add a flag (separate from the --apiservers flag) that skips the register node function but lets the kubelet still try to post node status updates if that will satisfy your ops folks.

smarterclayton commented:
----- Original Message -----
> So you want the kubelet to be able to connect to an apiserver (to send events
> and node status updates) but not to register itself (e.g. the old behavior)?
> And you'll use something like kube-register to manually create each of the
> nodes in the cluster? (Just trying to better understand your use case)

Essentially yes.

> 

> Accepting each kubelet before it joins the cluster is part of the design for
> dynamic clustering (post v1.0)  but is out of scope for static clustering
> (what this PR implements). Hopefully when we converge on dynamic clustering
> this becomes much simpler.
> 
> I can certainly add a flag (separate from the --apiservers flag) that skips
> the register node function but lets the kubelet still try to post node
> status updates if that will satisfy your ops folks.

I believe that would satisfy their use case of they retain control of node creation, but it's still acceptable for node/pod updates to come from a node.

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#issuecomment-97095373

bgrant0607 commented:
Kubelet definitely should handle rejection by apiserver gracefully. 

If we had a way of conveying different identities to different Kubelets or groups of Kubelets, then it seems like we could have a policy that determined whether they should be admitted. I could even imagine restricting scheduling based on such identities -- we call such a thing "dedicated machines". The system just has to trust that the node is reasonably well behaved. Users could decide which nodes they trusted to run their pods, and nodes could decide which users' pods they would accept.

gmarek commented:
I'd certainly want a test for registerNode() for existing node, as IIUC there was a bug there in the initial version (and for 'normal' case as well). 

Also it'd be nice to have a test for updating Node Address in case it's changed by a cloud provider, to make sure it solves #6072. But that's only a 'good to have'.

erictune commented:
Taking a look now.  Needs a rebase, BTW.

roberthbailey commented:
We were hoping to get #5246 in this week to unblock progress on a few other fronts which will require some re-work on this PR. But that one isn't yet passing e2e tests so this is on ice for a bit longer.

erictune commented:
Looks like these flags don't do anything:
--node-milli-cpu
--node-memory
--machines
--minion-regexp

Mention in description that these are dead, so that it gets into release notes.

Suggest deleting them from node-controller (will regen docs).
Suggest you grep for every use of these flags under docs/getting-started-guides and cluster and file an issue against the corresponding maintainer.  As the maintainer to delete the use of the flags, and retest with latest kubernetes, when they are ready. I expect that most of the providers will just work right with the flags deleted.

Delete these arguments to NewNodeController, and corresponding fields of NodeController struct.
matchRE string,
nodes []string,
staticResources *api.NodeResources,

Update references to removed flags in docs/node.md.  Actually, probably need to totally rewrite that doc.

erictune commented:
All those comments can be addressed in separate PR.

I'm out next week so assigning to CJ.

eparis commented:
If it's using spf13/pflag you can use FlagSet.MarkDeprecated(flag_name, error_message)

If you want them to still parse but warn the user....

eparis commented:
specifically: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/Godeps/_workspace/src/github.com/spf13/pflag/flag.go#L259

roberthbailey commented:
Rebased. Added a flag to suppress the new behavior for @smarterclayton. Filed #8286 to address @erictune's comments about cleaning up unused flags. Still need to add a couple of tests for @gmarek but other than that this is ready for another round of comments. I verified that `kube-up` still works so it is at least able to form clusters.

roberthbailey commented on pkg/kubelet/kubelet.go:
```diff
@@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error {
 	info, err := kl.GetCachedMachineInfo()
 	if err != nil {
-		glog.Errorf("error getting machine info: %v", err)
+		// TODO(roberthbailey): This is required for test-cmd.sh to pass.
```
@vishh @vmarmol Can you help answer the question I put in this TODO? Should I put a default capacity here (which is what test-cmd.sh expects) or should we not be picking a random capacity if cadvisor doesn't let us know what the node capacity is (e.g. update the test to stop checking for this)? It would be better if cadvisor just worked in test-cmd.sh but until then we need to decide which way to go here. Thanks.

davidopp replied:
IMO if cAdvisor is the only way for us to get capacity for the node, then we should report zero capacity until the information is available.

gmarek replied:
+1 to what @davidopp wrote.

vmarmol replied:
+1 to a zero capacity being a sane default. cAdvisor should work in test-cmd in Linux builds (unless we're injecting a fake).

smarterclayton replied:
Does this remove the ability of the administrator to set arbitrary capacity to a node via the API (if the kubelet always writes 0)?

----- Original Message -----
> > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error {
> >  	// cAdvisor locally, e.g. for test-cmd.sh, and in integration test.
> >  	info, err := kl.GetCachedMachineInfo()
> >  	if err != nil {
> > -		glog.Errorf("error getting machine info: %v", err)
> > +		// TODO(roberthbailey): This is required for test-cmd.sh to pass.
> 
> +1 to a zero capacity being a sane default. cAdvisor should work in test-cmd
> in Linux builds (unless we're injecting a fake).
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30424260

vmarmol replied:
@smarterclayton, that exists?

On Fri, May 15, 2015 at 9:51 AM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#discussion_r30425249>
> :
>
> > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error {
> >  	// cAdvisor locally, e.g. for test-cmd.sh, and in integration test.
> >  	info, err := kl.GetCachedMachineInfo()
> >  	if err != nil {
> > -		glog.Errorf("error getting machine info: %v", err)
> > +		// TODO(roberthbailey): This is required for test-cmd.sh to pass.
>
> Does this remove the ability of the administrator to set arbitrary
> capacity to a node via the API (if the kubelet always writes 0)?
> … <#14d587d3cb08fee9_>
> ----- Original Message -----
>  > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error
> { > // cAdvisor locally, e.g. for test-cmd.sh, and in integration test. >
> info, err := kl.GetCachedMachineInfo() > if err != nil { > -
> glog.Errorf("error getting machine info: %v", err) > + //
> TODO(roberthbailey): This is required for test-cmd.sh to pass. +1 to a zero
> capacity being a sane default. cAdvisor should work in test-cmd in Linux
> builds (unless we're injecting a fake). --- Reply to this email directly or
> view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30424260
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30425249>
> .
>

smarterclayton replied:
Anyone can call the node api.  If cadvisor can't report anything, and the node already has something, should the node overwrite the capacity?  I would argue it should set zero if nothing is set, but if it's explicitly set (perhaps from the last time) it shouldn't be reset.  Not a massively strong feeling, but if cadvisor is dead I don't know that we should be making the node completely unschedulable (by forcing capacity to 0)

vmarmol replied:
That sounds reasonable and it'll allow us to keep data we knew to be true.

It is important to note that under normal circumstances cAdvisor is
running, I have yet to see a situation in production where it is not. These
edge cases happen in testing where we mock or stub out different parts of
the Kubelet.

On Fri, May 15, 2015 at 10:08 AM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#discussion_r30426565>
> :
>
> > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error {
> >  	// cAdvisor locally, e.g. for test-cmd.sh, and in integration test.
> >  	info, err := kl.GetCachedMachineInfo()
> >  	if err != nil {
> > -		glog.Errorf("error getting machine info: %v", err)
> > +		// TODO(roberthbailey): This is required for test-cmd.sh to pass.
>
> Anyone can call the node api. If cadvisor can't report anything, and the
> node already has something, should the node overwrite the capacity? I would
> argue it should set zero if nothing is set, but if it's explicitly set
> (perhaps from the last time) it shouldn't be reset. Not a massively strong
> feeling, but if cadvisor is dead I don't know that we should be making the
> node completely unschedulable (by forcing capacity to 0)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30426565>
> .
>

smarterclayton replied:
If we ever split cadvisor out into its own process we'd potentially have problems with dead cadvisor.  Was there any plan for that as an option?

----- Original Message -----
> > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error {
> >  	// cAdvisor locally, e.g. for test-cmd.sh, and in integration test.
> >  	info, err := kl.GetCachedMachineInfo()
> >  	if err != nil {
> > -		glog.Errorf("error getting machine info: %v", err)
> > +		// TODO(roberthbailey): This is required for test-cmd.sh to pass.
> 
> That sounds reasonable and it'll allow us to keep data we knew to be true.
> 
> It is important to note that under normal circumstances cAdvisor is
> running, I have yet to see a situation in production where it is not. These
> edge cases happen in testing where we mock or stub out different parts of
> the Kubelet.
> 
> On Fri, May 15, 2015 at 10:08 AM, Clayton Coleman <notifications@github.com>
> wrote:
> 
> > In pkg/kubelet/kubelet.go
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#discussion_r30426565>
> > :
> >
> > > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error {
> > >  	// cAdvisor locally, e.g. for test-cmd.sh, and in integration test.
> > >  	info, err := kl.GetCachedMachineInfo()
> > >  	if err != nil {
> > > -		glog.Errorf("error getting machine info: %v", err)
> > > +		// TODO(roberthbailey): This is required for test-cmd.sh to pass.
> >
> > Anyone can call the node api. If cadvisor can't report anything, and the
> > node already has something, should the node overwrite the capacity? I would
> > argue it should set zero if nothing is set, but if it's explicitly set
> > (perhaps from the last time) it shouldn't be reset. Not a massively strong
> > feeling, but if cadvisor is dead I don't know that we should be making the
> > node completely unschedulable (by forcing capacity to 0)
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30426565>
> > .
> >
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30427038

vmarmol replied:
We never got that far AFAIK. Avoiding some of that was why we merged the
two.

On Fri, May 15, 2015 at 11:37 AM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#discussion_r30434044>
> :
>
> > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error {
> >  	// cAdvisor locally, e.g. for test-cmd.sh, and in integration test.
> >  	info, err := kl.GetCachedMachineInfo()
> >  	if err != nil {
> > -		glog.Errorf("error getting machine info: %v", err)
> > +		// TODO(roberthbailey): This is required for test-cmd.sh to pass.
>
> If we ever split cadvisor out into its own process we'd potentially have
> problems with dead cadvisor. Was there any plan for that as an option?
> … <#14d58ddd22a4fbf6_>
> ----- Original Message -----
>  > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus() error
> { > // cAdvisor locally, e.g. for test-cmd.sh, and in integration test. >
> info, err := kl.GetCachedMachineInfo() > if err != nil { > -
> glog.Errorf("error getting machine info: %v", err) > + //
> TODO(roberthbailey): This is required for test-cmd.sh to pass. That sounds
> reasonable and it'll allow us to keep data we knew to be true. It is
> important to note that under normal circumstances cAdvisor is running, I
> have yet to see a situation in production where it is not. These edge cases
> happen in testing where we mock or stub out different parts of the Kubelet.
> On Fri, May 15, 2015 at 10:08 AM, Clayton Coleman <
> notifications@github.com> wrote: > In pkg/kubelet/kubelet.go > <
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949#discussion_r30426565>
> > : > > > @@ -1668,11 +1753,18 @@ func (kl *Kubelet) tryUpdateNodeStatus()
> error { > > // cAdvisor locally, e.g. for test-cmd.sh, and in integration
> test. > > info, err := kl.GetCachedMachineInfo() > > if err != nil { > > -
> glog.Errorf("error getting machine info: %v", err) > > + //
> TODO(roberthbailey): This is required for test-cmd.sh to pass. > > Anyone
> can call the node api. If cadvisor can't report anything, and the > node
> already has something, should the node overwrite the capacity? I would >
> argue it should set zero if nothing is set, but if it's explicitly set >
> (perhaps from the last time) it shouldn't be reset. Not a massively strong
> > feeling, but if cadvisor is dead I don't know that we should be making
> the > node completely unschedulable (by forcing capacity to 0) > > — >
> Reply to this email directly or view it on GitHub > <
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30426565>
> > . > --- Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30427038
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6949/files#r30434044>
> .
>

davidopp replied:
If cAdvisor is integral to the system (which it seems it is), maybe we should refuse to register a node if cAdvisor is down (except in tests when it can't be run, I guess), and if it dies later for some amount of time we should report NodeReady==false?

roberthbailey replied:
Would folks be ok if we set the capacity to zero for the sake of this PR and filed an issue to continue the discussion?

roberthbailey replied:
Filed https://github.com/GoogleCloudPlatform/kubernetes/issues/8361

davidopp commented on pkg/kubelet/kubelet.go:
```diff
@@ -678,20 +680,83 @@ func (kl *Kubelet) runOOMWatcher() {
+					glog.Errorf("no node instance returned for %q", kl.hostname)
+					continue
+				}
+				if currentNode.Spec.ExternalID == node.Spec.ExternalID {
```
What does it mean if we get here and these are != ? Should we delete and re-register the node instead of just returning?

roberthbailey replied:
If they aren't equal we don't return, we loop forever. This check is basically making sure that both the node name and the external id match for the existing entry to be the same as this node.

roberthbailey commented:
@gmarek - Added a test for a node registering when it already exists (this should also be covered by the imminent reboot e2e test).

roberthbailey commented:
Re-pushed with my latest changes. PTAL.

roberthbailey commented:
Integration tests aren't currently passing. I'll investigate.

roberthbailey commented:
Turns out I forgot to set a single flag and now the integration tests are passing. While investigating, I started doing a bit of flag cleanup (first part of #8286) before finding the problem. 

Running e2e tests now (cluster creation succeeded, which bodes well).

@erictune @smarterclayton PTAL.

Commit:
Configure the cloud provider for the kubelet.

roberthbailey commented:
Just finished running e2e tests: 

```
Ran 43 of 49 Specs in 1678.866 seconds
SUCCESS! -- 43 Passed | 0 Failed | 1 Pending | 5 Skipped PASS
```

mbforbes commented:
(Needs rebase)

roberthbailey commented:
Rebased again.

erictune commented:
reviewing now

roberthbailey commented:
Thanks @erictune

erictune commented:
LGTM

mbforbes commented:
/xref should fix #8315 (linking for tracking)

lavalamp commented on pkg/cloudprovider/nodecontroller/nodecontroller.go:
```diff
@@ -156,37 +148,46 @@ func (nc *NodeController) generateCIDRs(num int) util.StringSet {
-		availableCIDRs.Delete(node.Spec.PodCIDR)
+// reconcilePodCIDRs looks at each node and assigns it a valid CIDR
+// if it doesn't currently have one.
+func (nc *NodeController) reconcilePodCIDRs(nodes *api.NodeList) {
```
Should it be called reconcile*Node*CIDRs?

roberthbailey replied:
Yes, probably should have renamed it while I was in here...

roberthbailey replied:
#8584

lavalamp commented on pkg/cloudprovider/nodecontroller/nodecontroller.go:
```diff
@@ -156,37 +148,46 @@ func (nc *NodeController) generateCIDRs(num int) util.StringSet {
-			podCIDR, _ = availableCIDRs.PopAny()
+	for _, node := range nodes.Items {
+		if node.Spec.PodCIDR == "" {
+			podCIDR, found := availableCIDRs.PopAny()
```
The CIDR has to match the network forwarding rules?

roberthbailey replied:
Yes, that's why we only persist it if nc.configureNodeCIDR returns success (b/c that's what configures the route on GCE).

lavalamp commented on pkg/cloudprovider/nodecontroller/nodecontroller.go:
```diff
@@ -156,37 +148,46 @@ func (nc *NodeController) generateCIDRs(num int) util.StringSet {
+				// The newly assigned CIDR was not properly configured, so don't save it in the API server.
+				continue
+			}
+			if _, err := nc.kubeClient.Nodes().Update(&node); err != nil {
```
This should be a get/update/retry loop. From my logs:

```
nodecontroller.go:179] Unable to assign node e2e-test-dbsmith-minion-f48d CIDR 10.245.0.0/24: minion "e2e-test-dbsmith-minion-f48d" cannot be upd
ated: the resource was updated to 154
E0520 18:43:19.622813       5 nodecontroller.go:179] Unable to assign node e2e-test-dbsmith-minion-nv8h CIDR 10.245.1.0/24: minion "e2e-test-dbsmith-minion-nv8h" cannot be upd
ated: the resource was updated to 192
```

roberthbailey replied:
reconcilePodCIDRs is called every N seconds. So it will retry on the next loop.
